plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java

```diff
@@ -0,0 +1,225 @@
+import java.util.concurrent.CompletableFuture;
+
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Collections.emptyIterator;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Concatenates the Pages for all PageSources which were grouped together an Iceberg CombinedScanTask.
+ */
+public class IcebergCombinedPageSource
+        implements UpdatablePageSource
+{
+    private final Map<String, UpdatablePageSource> delegatePageSources;
+    private final Iterator<UpdatablePageSource> pageSourceIterator;
+
+    private ConnectorPageSource currentPageSource;
+
+    public IcebergCombinedPageSource(Map<String, UpdatablePageSource> delegatePageSourcesByPath)
+    {
+        if (delegatePageSourcesByPath.isEmpty()) {
+            this.delegatePageSources = ImmutableMap.of();
+            this.pageSourceIterator = emptyIterator();
+            this.currentPageSource = new EmptyPageSource();
+        }
+        else {
+            this.delegatePageSources = ImmutableMap.copyOf(requireNonNull(delegatePageSourcesByPath, "delegatePageSources is null"));
+            this.pageSourceIterator = delegatePageSourcesByPath.values().iterator();
+            this.currentPageSource = pageSourceIterator.next();
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegatePageSources.values().stream()
+                .mapToLong(ConnectorPageSource::getCompletedBytes)
+                .sum();
+    }
+
+    @Override
+    public OptionalLong getCompletedPositions()
+    {
+        return delegatePageSources.values().stream()
+                .map(ConnectorPageSource::getCompletedPositions)
+                .reduce(OptionalLong.of(0), (accumulatedPositions, newPositions) -> {
+                    if (accumulatedPositions.isPresent() && newPositions.isPresent()) {
+                        return OptionalLong.of(accumulatedPositions.getAsLong() + newPositions.getAsLong());
+                    }
+                    return OptionalLong.empty();
+                });
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return delegatePageSources.values().stream()
+                .mapToLong(ConnectorPageSource::getReadTimeNanos)
+                .sum();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return !pageSourceIterator.hasNext() && currentPageSource.isFinished();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        if (currentPageSource.isFinished()) {
+            currentPageSource = pageSourceIterator.next();
+        }
+
+        return currentPageSource.getNextPage();
+    }
+
+    @Override
+    public long getMemoryUsage()
+    {
+        return delegatePageSources.values().stream()
+                .mapToLong(ConnectorPageSource::getMemoryUsage)
+                .sum();
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        IOException closeException = null;
```

Comment:
Simply add them all to a `com.google.common.io.Closer` which handles this automatically.